### PR TITLE
Skip the remaining callbacks on connectivity state change

### PR DESCRIPTION
**Pull request by Hexta**
Skip the remaining callbacks in `subchannel.transitionToState()` if connectivity state has been changed before callbacks called.

Connectivity state could be changed by calling `client.close()` inside a callback.
As a result - infinite busy loop in `LoadBalancingCall.doPick()`.

It affected [etcd3 client library](https://github.com/microsoft/etcd3/blob/00bfb8bf3a26e176c48b2f483d70df0605a76006/src/connection-pool.ts#L213]).

### CODE REVIEW

**Comment by murgatroid99** [comment]
The change seems reasonable, but I don't understand the test. Why should `client2.waitForReady` end with an error? Why should `client2` then be able to make a request successfully?

**Comment by Hexta** [comment]
`client2.waitForReady` will end with `Failed to connect before the deadline` due to skipped the remaining connectivity state listeners. The subchannel will remain in `IDLE` state.
On the next request the subchannel will transit to `CONNECTING` state due to call stack:
- `Subchannel.transitionToState`
- `Subchannel.startConnecting`
- `ChannelSubchannelWrapper.startConnecting`
- `PickFirstLoadBalancer.exitIdle`
- `ChildLoadBalancerHandler.exitIdle`
- `ResolvingLoadBalancer.exitIdle`
- `InternalChannel.getConfig`
- `ResolvingCall.getConfig`
- `ResolvingCall.start`
- `BaseUnaryInterceptingCall.start`
- `BaseUnaryInterceptingCall.start`
- `ServiceClientImpl.makeUnaryRequest`

**Comment by Hexta** [comment]
Unfortunately, `client1.close` affects the subchannel's connectivity state.
The subchannel will remain in `IDLE` state even without this change, right.

In the case of 2 clients, there are 4 registered connectivity state listeners in the beginning of transition `CONNECTING` -> `READY`:

1. `ChannelSubchannelWrapper.subchannelStateListener` for 1st channel
2. `PickFirstLoadBalancer.pickedSubchannelStateListener` for 1st channel
3. `ChannelSubchannelWrapper.subchannelStateListener` for 2nd channel
4. `PickFirstLoadBalancer.pickedSubchannelStateListener` for 2nd channel

2nd listener calls `client1.close`.
<details>
<summary>call stack  (line numbers could be a little bit different due to some debug prints)</summary>
- `/grpc-js/test/test-global-subchannel-pool.ts:97:55`
- `Object.checkState [as callback] (/grpc-js/src/client.ts:188:9)`
- `InternalChannel.updateState (/grpc-js/src/internal-channel.ts:397:23)`
- `Object.updateState (/grpc-js/src/internal-channel.ts:250:14)`
- `ResolvingLoadBalancer.updateState (/grpc-js/src/resolving-load-balancer.ts:288:31)`
- `Object.updateState (/grpc-js/src/resolving-load-balancer.ts:173:14)`
- `ChildLoadBalancerHandler.ChildPolicyHelper.updateState (/grpc-js/src/load-balancer-child-handler.ts:61:40)`
- `PickFirstLoadBalancer.updateState (/grpc-js/src/load-balancer-pick-first.ts:335:31)`
- `PickFirstLoadBalancer.pickSubchannel (/grpc-js/src/load-balancer-pick-first.ts:320:10)`
- `PickFirstLoadBalancer.subchannelStateListener (/grpc-js/src/load-balancer-pick-first.ts:169:14)`
- `Subchannel.transitionToState (/grpc-js/src/subchannel.ts:272:7)`
- `/grpc-js/src/subchannel.ts:179:18`
</details>

`client1.close` changes the subchannel's connectivity state `READY` -> `IDLE`.
<details>
<summary>call stack</summary>
- `Subchannel.transitionToState (/grpc-js/src/subchannel.ts:228:22)`
- `Subchannel.unref (/grpc-js/src/subchannel.ts:333:12)`
- `Subchannel.unrefIfOneRef (/grpc-js/src/subchannel.ts:345:12)`
- `/grpc-js/src/subchannel-pool.ts:73:38`
- `Array.filter (<anonymous>)`
- `SubchannelPool.unrefUnusedSubchannels (/grpc-js/src/subchannel-pool.ts:72:51)`
- `InternalChannel.close (/grpc-js/src/internal-channel.ts:575:25)`
- `ChannelImplementation.close (/grpc-js/src/channel.ts:120:26)`
- `/grpc-js/test/test-global-subchannel-pool.ts:139:30`
- `Object.checkState [as callback] (/grpc-js/src/client.ts:189:9)`
- `InternalChannel.updateState (/grpc-js/src/internal-channel.ts:421:23)`
- `Object.updateState (/grpc-js/src/internal-channel.ts:276:14)`
- `ResolvingLoadBalancer.updateState (/grpc-js/src/resolving-load-balancer.ts:289:31)`
- `Object.updateState (/grpc-js/src/resolving-load-balancer.ts:173:14)`
- `ChildLoadBalancerHandler.ChildPolicyHelper.updateState (/grpc-js/src/load-balancer-child-handler.ts:60:40)`
- `PickFirstLoadBalancer.updateState (/grpc-js/src/load-balancer-pick-first.ts:335:31)`
- `PickFirstLoadBalancer.pickSubchannel (/grpc-js/src/load-balancer-pick-first.ts:320:10)`
- `PickFirstLoadBalancer.subchannelStateListener (/grpc-js/src/load-balancer-pick-first.ts:169:14)`
- `Subchannel.transitionToState (/grpc-js/src/subchannel.ts:300:7)`
- `/grpc-js/src/subchannel.ts:184:18`
</details>

`Subchannel.unrefIfOneRef` is called because `refcount === 1` at this moment because 
`PickFirstLoadBalancer.resetSubchannelList` called `Subchannel.unref` twice on `client1.close`.

<details>
<summary>1st attempt</summary>
- `Subchannel.unref (/grpc-js/src/subchannel.ts:292:22)`
- `ChannelSubchannelWrapper.unref (/grpc-js/src/internal-channel.ts:105:16)`
- `PickFirstLoadBalancer.resetSubchannelList (/grpc-js/src/load-balancer-pick-first.ts:341:18)`
- `PickFirstLoadBalancer.destroy (/grpc-js/src/load-balancer-pick-first.ts:451:10)`
- `ChildLoadBalancerHandler.destroy (/grpc-js/src/load-balancer-child-handler.ts:151:25)`
- `ResolvingLoadBalancer.destroy (/grpc-js/src/resolving-load-balancer.ts:325:28)`
- `InternalChannel.close (/grpc-js/src/internal-channel.ts:544:32)`
- `ChannelImplementation.close (/grpc-js/src/channel.ts:120:26)`
- `ServiceClientImpl.close (/grpc-js/src/client.ts:167:26)`
- `/grpc-js/test/test-global-subchannel-pool.ts:98:20`
- `Object.checkState [as callback] (/grpc-js/src/client.ts:188:9)`
- `InternalChannel.updateState (/grpc-js/src/internal-channel.ts:397:23)`
- `Object.updateState (/grpc-js/src/internal-channel.ts:250:14)`
- `ResolvingLoadBalancer.updateState (/grpc-js/src/resolving-load-balancer.ts:288:31)`
- `Object.updateState (/grpc-js/src/resolving-load-balancer.ts:173:14)`
- `ChildLoadBalancerHandler.ChildPolicyHelper.updateState (/grpc-js/src/load-balancer-child-handler.ts:61:40)`
- `PickFirstLoadBalancer.updateState (/grpc-js/src/load-balancer-pick-first.ts:335:31)`
- `PickFirstLoadBalancer.pickSubchannel (/grpc-js/src/load-balancer-pick-first.ts:320:10)`
- `PickFirstLoadBalancer.subchannelStateListener (/grpc-js/src/load-balancer-pick-first.ts:169:14)`
- `Subchannel.transitionToState (/grpc-js/src/subchannel.ts:272:7)`
- `/grpc-js/src/subchannel.ts:179:18`
</details>

<details>
<summary>2nd attempt</summary>
- `Subchannel.unref (/grpc-js/src/subchannel.ts:292:22)`
- `PickFirstLoadBalancer.destroy (/grpc-js/src/load-balancer-pick-first.ts:457:19)`
- `ChildLoadBalancerHandler.destroy (/grpc-js/src/load-balancer-child-handler.ts:151:25)`
- `ResolvingLoadBalancer.destroy (/grpc-js/src/resolving-load-balancer.ts:325:28)`
- `InternalChannel.close (/grpc-js/src/internal-channel.ts:544:32)`
- `ChannelImplementation.close (/grpc-js/src/channel.ts:120:26)`
- `ServiceClientImpl.close (/grpc-js/src/client.ts:167:26)`
- `/grpc-js/test/test-global-subchannel-pool.ts:98:20`
- `Object.checkState [as callback] (/grpc-js/src/client.ts:188:9)`
- `InternalChannel.updateState (/grpc-js/src/internal-channel.ts:397:23)`
- `Object.updateState (/grpc-js/src/internal-channel.ts:250:14)`
- `ResolvingLoadBalancer.updateState (/grpc-js/src/resolving-load-balancer.ts:288:31)`
- `Object.updateState (/grpc-js/src/resolving-load-balancer.ts:173:14)`
- `ChildLoadBalancerHandler.ChildPolicyHelper.updateState (/grpc-js/src/load-balancer-child-handler.ts:61:40)`
- `PickFirstLoadBalancer.updateState (/grpc-js/src/load-balancer-pick-first.ts:335:31)`
- `PickFirstLoadBalancer.pickSubchannel (/grpc-js/src/load-balancer-pick-first.ts:320:10)`
- `PickFirstLoadBalancer.subchannelStateListener (/grpc-js/src/load-balancer-pick-first.ts:169:14)`
- `Subchannel.transitionToState (/grpc-js/src/subchannel.ts:272:7)`
- `/grpc-js/src/subchannel.ts:179:18`
- `processTicksAndRejections (node:internal/process/task_queues:95:5)`
</details>

`PickFirstLoadBalancer.destroy` calls
1. `PickFirstLoadBalancer.resetSubchannelList()` 
1. `PickFirstLoadBalancer.currentPick.unref()`

`PickFirstLoadBalancer.resetSubchannelList()` calls `.unref()` for every subchannel in `PickFirstLoadBalancer.subchannels`.

Array `PickFirstLoadBalancer.subchannels` contains the subchannel when 2nd listener called.
The subchannel is still in the array because 2nd listener (`PickFirstLoadBalancer.subchannelStateListener`) calls `this.updateState()` before calling `this.resetSubchannelList()`.

TL;DR
I don’t t mind calling the listener with `this.connectivityState` instead of `newState`.
I would like to note that it'd not change the behaviour of 1st attempt of `client2.waitForReady()`.
There will be still the same error `Deadline passed without connectivity state change` because `client2` will remain in `CONNECTING` state.

**Comment by murgatroid99** [comment]
> The subchannel is still in the array because 2nd listener (`PickFirstLoadBalancer.subchannelStateListener`) calls `this.updateState()` before calling `this.resetSubchannelList()`.

This is the bug. And actually, when I fix that, there is another bug elsewhere.

I now believe that the change you have here is not what is needed. I'll put together a separate PR.

**Comment by murgatroid99** [comment]
I published version 1.8.14 with a change that seems to fix this problem. Please try it out.
